core/src/main/java/gyro/core/scope/RootScope.java

```diff
@@ -253,6 +254,43 @@ public Object convertValue(Type returnType, Object object) {
         return stream.collect(Collectors.toList());
     }
 
+    public List<Resource> findSortedResourcesIn(Set<String> diffFiles) {
+        List<Resource> allResources = findResourcesIn(diffFiles);
+        List<Resource> sortedResources = new ArrayList<>();
+        Defer.execute(allResources, i -> addResource(i, sortedResources, allResources));
+        return sortedResources;
+    }
+
+    private void addResource(Resource resource, List<Resource> sortedResources, List<Resource> allResources) {
+        if (shouldAddResource(resource, sortedResources, allResources)) {
+            sortedResources.add(resource);
+        } else {
+            throw new Defer(null, String.format("Can't resolve dependencies of @|bold %s|@!", resource.primaryKey()));
+        }
+    }
+
+    private boolean shouldAddResource(Diffable diffable, List<Resource> sortedResources, List<Resource> allResources) {
+        for (DiffableField field : DiffableType.getInstance(diffable.getClass()).getFields()) {
+            Object value = field.getValue(diffable);
+
+            for (Object v : (value instanceof Collection
+                ? ((Collection<?>) value)
+                : Collections.singletonList(value))) {
+                if (v instanceof Resource) {
+                    if (allResources.contains(v) && !sortedResources.contains(v)) {
+                        return false;
+                    }
+                } else if (v instanceof Diffable) {
+                    if (!shouldAddResource((Diffable) v, sortedResources, allResources)) {
+                        return false;
+                    }
+                }
```

Comment:
Can this be simplified like below ?

```suggestion
                if (v instanceof Resource && allResources.contains(v) && !sortedResources.contains(v)) {
                    return false;
                } else if (v instanceof Diffable && !shouldAddResource((Diffable) v, sortedResources, allResources)) {
                    return false;
                }
```

Reply:
I fixed the `else if`, but the first check should stay the same because we don't want to call `shouldAddResource` on Resources.